artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerSessionImpl.java

```diff
@@ -1789,6 +1790,11 @@ public synchronized RoutingStatus doSend(final Transaction tx,
 
       AddressInfo art = getAddressAndRoutingType(new AddressInfo(msg.getAddressSimpleString(), routingType));
 
+      if (postOffice.isAddressBlocked(msg.getAddressSimpleString())) {
```

Comment:
I think the use case here is wrong. With block you just stop sending credits to the client.

This becomes fail at this point.

We have two options:


- If you had previously sent credits to the client.. then you must complete the operations until you have full credits.

- or we could hung to these on the server. But that would have a risk of running OME. So I would say we just stop sending credits.

Reply:
@clebertsuconic reply here.
I can think of 2 difficulties using credits:
1. Some clients (like mqtt, stomp) are not supporting flow control so they can't be blocked using credits.
2. producers are not necessarily bound to one address, they can send messages to different addresses (for example one producer can send one message to address1, and the next one to address2. It it's blocked on address1, the next message also be blocked)

If we simply block the sending threads, it will holding up broker threads (like you said OME). Besides the producer may have call-timeout exception.

By throwing a proper exception we can let client to handle their producer properly without blocking any thing.

Reply:
btw my test missed stomp and mqtt, I'll add them.

Reply:
We should have discussed this before you coming into implementation. 

I don't think this is the right design.

Reply:
You mean a formal discussion? I'll send an email for that purpose. I think it not too late. :)